src/serac/physics/materials/hyperelastic_material.cpp

```diff
@@ -0,0 +1,175 @@
+  double dJ = F.Det();
+
+  double a = mu_ * std::pow(dJ, -(2.0 / dim) - 1.0);
+  double b = bulk_ * (dJ - 1.0) - a * (F * F) / dim;
+
+  mfem::MultABt(F, F, B_);
+
+  sigma = 0.0;
+
+  sigma.Add(a, B_);
+
+  for (int i = 0; i < dim; ++i) {
+    sigma(i, i) += b;
+  }
+}
+
+void NeoHookeanMaterial::AssembleTangentModuli(const mfem::DenseMatrix& F, mfem_ext::Array4D<double>& C) const
+{
+  int dim = F.Width();
+  B_.SetSize(dim);
+  C.SetSize(dim, dim, dim, dim);
+
+  mfem::MultABt(F, F, B_);
+
+  double dJ = F.Det();
+
+  if (c_mu_) {
+    EvalCoeffs();
+  }
+
+  // See http://solidmechanics.org/Text/Chapter8_4/Chapter8_4.php for a sample derivation
+  double a = mu_ * std::pow(dJ, -2.0 / dim);
+  double b = bulk_ * (2.0 * dJ - 1.0) * dJ + a * (2.0 / (dim * dim)) * (F * F);
+
+  C = 0.0;
+
+  for (int i = 0; i < dim; ++i) {
+    for (int j = 0; j < dim; ++j) {
+      for (int k = 0; k < dim; ++k) {
+        for (int l = 0; l < dim; ++l) {
+          if (i == k) {
+            C(i, j, k, l) += a * B_(j, l);
+          }
+          if (j == k) {
+            C(i, j, k, l) += a * B_(i, l);
+          }
+          if (k == l) {
+            C(i, j, k, l) -= a * (2.0 / dim) * B_(i, j);
+          }
+          if (i == j) {
+            C(i, j, k, l) -= a * (2.0 / dim) * B_(k, l);
+          }
+          if (i == j && k == l) {
+            C(i, j, k, l) += b;
+          }
+        }
+      }
+    }
+  }
+}
+
+inline void LinearElasticMaterial::EvalCoeffs() const
+{
+  mu_   = c_mu_->Eval(*Ttr_, Ttr_->GetIntPoint());
+  bulk_ = c_bulk_->Eval(*Ttr_, Ttr_->GetIntPoint());
+}
+
+void LinearElasticMaterial::EvalStress(const mfem::DenseMatrix& F, mfem::DenseMatrix& sigma) const
+{
+  int dim = F.Width();
+  sigma.SetSize(dim);
+  epsilon_.SetSize(dim);
+  FT_.SetSize(dim);
+
+  if (c_mu_) {
+    EvalCoeffs();
+  }
+
+  // Evaluate the linearized strain tensor from the deformation gradient
+  epsilon_ = 0.0;
+  epsilon_.Add(0.5, F);
+  FT_.Transpose(F);
+  epsilon_.Add(0.5, FT_);
+
+  for (int i = 0; i < dim; ++i) {
+    epsilon_(i, i) -= 1.0;
+  }
+
+  sigma                = 0.0;
+  double trace_epsilon = epsilon_.Trace();
+
+  // Calculate the stress by Hooke's law
+  sigma.Add(2.0 * mu_, epsilon_);
+  for (int i = 0; i < dim; ++i) {
+    sigma(i, i) += bulk_ * trace_epsilon - (2.0 / dim) * mu_ * trace_epsilon;
+  }
+}
+
+void LinearElasticMaterial::AssembleTangentModuli(const mfem::DenseMatrix& F, mfem_ext::Array4D<double>& C) const
+{
+  int dim = F.Width();
+
+  if (c_mu_) {
+    EvalCoeffs();
+  }
+
+  C = 0.0;
+
+  for (int i = 0; i < dim; ++i) {
+    for (int j = 0; j < dim; ++j) {
+      for (int k = 0; k < dim; ++k) {
+        for (int l = 0; l < dim; ++l) {
+          if (i == j && k == l) {
+            C(i, j, k, l) += bulk_ - (2.0 / dim) * mu_;
+          }
+          if (i == k && j == l) {
+            C(i, j, k, l) += mu_;
+          }
+          if (i == l && j == k) {
+            C(i, j, k, l) += mu_;
+          }
```

Comment:
this can be done branchlessly by observing that `int(i==j)` is the same as `kronecker_delta(i,j)`, so you get
```
C(i,j,k,l) += lambda * (i==j) * (k==l) + mu * ((i==k) * (j==l)  + (i==l) * (j==k));
```

Reply:
it's interesting to look at the generated assembly for different ways of populating `C`, here
https://godbolt.org/z/1oW674

1. your original approach generates a lot of branching for control flow and accumulation
2. the snippet above has a few branches for loop control flow (I think), but is pretty well inlined
3. a for_constexpr produces zero branching and 81 assignments, as the loop indices are evaluated at compile time
4. for_constexpr + `if constexpr` has zero branching and actually only sparsely updates the nonzero components!

Reply:
isn't better really to use the 6x6 form?

Reply:
Yes - I originally did this but had a typo somewhere so I reverted to a way that was easier to eyeball debug.

Reply:
> 
> 
> it's interesting to look at the generated assembly for different ways of populating `C`, here
> https://godbolt.org/z/1oW674
> 
>     1. your original approach generates a lot of branching for control flow and accumulation
> 
>     2. the snippet above has a few branches for loop control flow (I think), but is pretty well inlined
> 
>     3. a for_constexpr produces zero branching and 81 assignments, as the loop indices are evaluated at compile time
> 
>     4. for_constexpr + `if constexpr` has zero branching and actually only sparsely updates the nonzero components!

Awesome! I figured there were a bunch of ways to optimize this really dumb implementation.

Reply:
> isn't better really to use the 6x6 form?

I'm sure it is more efficient from a computational standpoint, but Voigt notation always seems to cause bugs due to the factor of 2 on the shear terms. I'd like to avoid it, if possible.